### DISCUSSION TITLE
[@types/react] BaseSyntheticEvent generic E is always Event

### DISCUSSION THREAD

**Discussion by nparoski** [DefinitelyTyped/DefinitelyTyped#58539] [Issues with a @types package]
I have issue with `SyntheticEvent` generic `E = Event` because it extends `BaseSyntheticEvent` interface and generic `E` is always `Event`. Essentially this means that `nativeEvent` is always going to be `Event` when in fact it can be `InputEvent`.

In my case I need to detect what type of inputEvent was triggered on that element ([list of inputEvents](https://rawgit.com/w3c/input-events/v1/index.html#interface-InputEvent-Attributes))

Consider next scenario:
```
const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
 console.log(e.nativeEvent.inputType)
}
```
There is error message `Property 'inputType' does not exist on type 'Event'.ts(2339)`

[a comment by an automated account is not shown]

**Comment by eps1lon** [comment]
Note that React makes no guarantees about what type `nativeEvent` is. Depending on the React event, user input and user agent it may be different event types. 

This is why we don't type them since these are effectively React implementation details.

**Reply by nparoski** [reply]
Hm okay, but this doesn't feel right, there is no way for me to override it essentially I am stuck with wrong type 🤔

**Reply by nparoski** [reply]
Thinking further about this, I do get it that they can't predict that but shouldn't it be typed differently or allow for some kind of override? I could live with having to use type guards to assure that that property indeed exists. Even though it may be more inconvenient  when working with it makes more sense to me than having it wrongly typed. If I am wrong please correct me

**Reply by ericanderson** [reply]
Put up a pr to add an optional generic to pipe through the event. That seems reasonable.

**Reply by nparoski** [reply]
Alright, I'll make PR by the end of the day.

**Reply by eps1lon** [reply]
Note that this will end up similarly to `Event.target`. We'll never be able to statically verify that the `nativeEvent` type matches so it's better to just defer to type-casting.

**Reply by ericanderson** [reply]
Agreed, it's not ideal at all. That said, this short hand is used in many places in DT. And it would be better for us to debate on an actual change.
